src/plugins/procdump/procdump.cpp

```diff
@@ -0,0 +1,777 @@
+                delete[] buf;
+        }
+    }
+
+    PRINT_DEBUG("[PID:%d] [%s] MMVAD 0x%13lx, 0x%13lx (%5lu PTEs, "
+                "%10zu bytes), MemCommit %d, VadType %d, CommitCharge 0x%lx\n",
+                ctx->pid, ctx->name.data(), mmvad->starting_vpn,
+                mmvad->ending_vpn, len_pages, len_bytes, vad_commit, vad_type,
+                vad_commit_charge);
+
+    ctx->vads[vad_start] = {vad_type, len_pages, prototype_pte, 0};
+    ctx->size += len_bytes;
+
+    return false;
+}
+
+static event_response_t terminate_process_cb(drakvuf_t drakvuf,
+                                             drakvuf_trap_info_t* info)
+{
+    attached_proc_data_t proc(drakvuf, info);
+    if (!proc.pid)
+    {
+        PRINT_DEBUG("[PROCDUMP] [PID:%d] [TID:%d] Error: Failed to get "
+                    "attached process\n",
+                    info->proc_data.pid, info->proc_data.tid);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    auto plugin = get_trap_plugin<procdump>(info);
+    if (!plugin)
+    {
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    // The callback could be called if other thread invokes NtTerminateProcess
+    // or as a return path from injected function.
+    // In both cases we should not starting process dump again.
+    auto it = plugin->terminating.find(proc.pid);
+    if (it != plugin->terminating.end())
+    {
+        if (!it->second)
+            plugin->terminating.erase(proc.pid);
+
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+    else
+        plugin->terminating[proc.pid] = proc.tid;
+
+    // TODO Move into constructor
+    auto ctx = new procdump_ctx;
+    ctx->pid = proc.pid;
+    ctx->ppid = proc.ppid;
+    ctx->tid = proc.tid;
+    ctx->name = std::string(proc.name);
+    ctx->plugin = plugin;
+    ctx->idx = plugin->procdumps_count++;
+    ctx->size = 0;
+    // Get virtual address space map of the process
+    drakvuf_traverse_mmvad(drakvuf, proc.base_addr, dump_mmvad,
+                           ctx);
+    if (ctx->vads.empty())
+    {
+        // nothing to do
+        delete ctx;
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+    {
+        PRINT_DEBUG("[PROCDUMP] [\"%s\":%4d] [PID:%4d] [TID:%4d] [\"%s\"] "
+                    "Dump 0x%lx (%ld MiB, %lu VADs) SN=%lu\n",
+                    __FUNCTION__, __LINE__, proc.pid,
+                    proc.tid, proc.name, ctx->size,
+                    ctx->size / 1024 / 1024, ctx->vads.size(), ctx->idx);
+    }
+    ctx->file_path =
+        plugin->procdump_dir + "/procdump." + std::to_string(ctx->idx);
+    ctx->file =
+        fopen((ctx->file_path + ".mm").data(), "w");
+    if (!ctx->file)
+    {
+        PRINT_DEBUG("[PROCDUMP] Failed to create file\n");
+        delete ctx;
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    // Save registers to restore process/thread state
+    memcpy(&ctx->saved_regs, info->regs, sizeof(x86_registers_t));
+    ctx->bp = (drakvuf_trap_t*)g_try_malloc0(sizeof(drakvuf_trap_t));
```

Comment:
Use `g_slice_new0` please

Reply:
Also, you need to keep track of all dynamically allocated traps in a separate list, you can't rely on it getting freed in their callback. That callback may never happen and you still need to be able to clean up the memory.

Reply:
Done